test_conformance/extensions/cl_khr_command_buffer/negative_command_nd_range_kernel.cpp

```diff
@@ -0,0 +1,535 @@
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_VALUE if values specified in properties are not valid.
+struct CommandNDRangeKernelInvalidProperties : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    cl_int Run() override
+    {
+        cl_ndrange_kernel_command_properties_khr empty_properties;
+
+        cl_int error = clCommandNDRangeKernelKHR(
+            command_buffer, nullptr, &empty_properties, kernel, 1, nullptr,
+            &num_elements, nullptr, 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_VALUE,
+                               "clCommandNDRangeKernelKHR should return "
+                               "CL_INVALID_VALUE",
+                               TEST_FAIL);
+
+        cl_ndrange_kernel_command_properties_khr props_invalid[3] = {
+            CL_MUTABLE_DISPATCH_UPDATABLE_FIELDS_KHR,
+            CL_MEM_USE_CACHED_CPU_MEMORY_IMG, 1
+        };
+
+        error = clCommandNDRangeKernelKHR(
+            command_buffer, nullptr, props_invalid, kernel, 1, nullptr,
+            &num_elements, nullptr, 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_VALUE,
+                               "clCommandNDRangeKernelKHR should return "
+                               "CL_INVALID_VALUE",
+                               TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_OPERATION if command_buffer has been finalized.
+struct CommandNDRangeKernelCommandBufferFinalized
+    : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    cl_int Run() override
+    {
+        cl_int error = clFinalizeCommandBufferKHR(command_buffer);
+        test_error(error, "clFinalizeCommandBufferKHR failed");
+
+        error = clCommandNDRangeKernelKHR(
+            command_buffer, nullptr, nullptr, kernel, 1, nullptr, &num_elements,
+            nullptr, 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_OPERATION,
+                               "clCommandNDRangeKernelKHR should return "
+                               "CL_INVALID_OPERATION",
+                               TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_VALUE if mutable_handle is not NULL.
+struct CommandNDRangeKernelMutableHandleNotNull : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    cl_int Run() override
+    {
+        cl_mutable_command_khr mutable_handle;
+
+        cl_int error = clCommandNDRangeKernelKHR(
+            command_buffer, nullptr, nullptr, kernel, 1, nullptr, &num_elements,
+            nullptr, 0, nullptr, nullptr, &mutable_handle);
+
+        test_failure_error_ret(error, CL_INVALID_VALUE,
+                               "clCommandNDRangeKernelKHR should return "
+                               "CL_INVALID_VALUE",
+                               TEST_FAIL);
+
```

Comment:
We should skip this test if `cl_khr_command_buffer_mutable_dispatch` is supported, as it relaxes this restriction.